File: app/src/test/java/com/maximillionsnyder/umafinidad/RankingPadresTest.kt
package com.maximillionsnyder.umafinidad

import com.maximillionsnyder.umafinidad.data.CharacterDto
import com.maximillionsnyder.umafinidad.data.MemberDto
import com.maximillionsnyder.umafinidad.data.RelationDto
import com.maximillionsnyder.umafinidad.data.jsonParser
import com.maximillionsnyder.umafinidad.data.toDomain
import com.maximillionsnyder.umafinidad.domain.AffinityModel
import kotlinx.serialization.decodeFromString
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

/* Ranking “Mejores padres”: cada hijo aporta su linaje óptimo exacto y
   cada uma cuenta cuántas veces sale como padre/madre. */
class RankingPadresTest {

    private lateinit var modelo: AffinityModel
    private lateinit var ids: List<Int>

    private fun recurso(ruta: String): String =
        javaClass.getResourceAsStream("/$ruta")!!.bufferedReader().use { it.readText() }

    @Before
    fun preparar() {
        val characters = jsonParser.decodeFromString<List<CharacterDto>>(recurso("data/characters.json"))
        val relations = jsonParser.decodeFromString<List<RelationDto>>(recurso("data/succession_relation.json"))
        val members = jsonParser.decodeFromString<List<MemberDto>>(recurso("data/succession_relation_member.json"))
        modelo = AffinityModel(
            characters.map { it.toDomain() },
            relations.map { it.toDomain() },
            members.map { it.toDomain() },
        )
        ids = characters.map { it.toDomain() }
            .filter { it.playable == true && it.active == true }
            .map { it.charId }
    }

    @Test
    fun cubreTodoElPool() {
        val ranking = modelo.rankingPadres()
        assertEquals(ids.size, ranking.size)
        assertEquals(ids.sorted(), ranking.map { it.personaje.charId }.sorted())
    }

    @Test
    fun cadaLinajeAportaDosPadres() {
        val ranking = modelo.rankingPadres()
        // Cada hijo con linaje computable aporta exactamente padre + madre.
        assertEquals(2 * ids.size, ranking.sumOf { it.veces })
    }

    @Test
    fun porcentajeCoherenteConVeces() {
        val ranking = modelo.rankingPadres()
        for (entry in ranking) {
            assertEquals(entry.veces * 100f / ids.size, entry.porcentaje, 0.001f)
            assertTrue("veces de ${entry.personaje.charId} en rango", entry.veces in 0 until ids.size)
            assertTrue(entry.puntosMedios >= 0)
        }
    }

    @Test
    fun ordenDeterminista() {
        val primero = modelo.rankingPadres()
        val segundo = modelo.rankingPadres()
        assertEquals(primero, segundo)
        assertTrue(primero.zipWithNext().all { (a, b) ->
            if (a.veces != b.veces) a.veces > b.veces
            else if (a.puntosMedios != b.puntosMedios) a.puntosMedios > b.puntosMedios
            else if (a.totalAfinidad != b.totalAfinidad) a.totalAfinidad > b.totalAfinidad
            else a.personaje.charId <= b.personaje.charId
        })
    }

    @Test
    fun recuentoIndependienteCoincide() {
        // Recuenta desde fuera con mejorLinajeDe y compara con el ranking.
        val conteo = HashMap<Int, Int>()
        val puntos = HashMap<Int, MutableList<Int>>()
        for (hijoId in ids) {
            val linaje = modelo.mejorLinajeDe(hijoId) ?: continue
            for (padreId in listOf(linaje.padre.charId, linaje.madre.charId)) {
                conteo[padreId] = (conteo[padreId] ?: 0) + 1
                puntos.getOrPut(padreId) { mutableListOf() }.add(linaje.puntos)
            }
        }
        val ranking = modelo.rankingPadres().associateBy { it.personaje.charId }
        for (id in ids) {
            val entry = ranking.getValue(id)
            assertEquals("veces de $id", conteo[id] ?: 0, entry.veces)
            val esperadaMedia = puntos[id]?.let { it.sum() / it.size } ?: 0
            assertEquals("media de $id", esperadaMedia, entry.puntosMedios)
        }
    }
}

File: sitio/src/lib/domain/__tests__/rankingPadres.test.ts
/* Ranking “Mejores padres”. Espejo de RankingPadresTest.kt. */

import { describe, expect, it } from 'vitest'
import { ids, modelo } from '../../../test/data'

const m = modelo()

describe('rankingPadres', () => {
  it('cubre todo el pool', () => {
    const ranking = m.rankingPadres()
    expect(ranking.length).toBe(ids.length)
    expect(ranking.map((e) => e.personaje.charId).sort((a, b) => a - b)).toEqual(
      [...ids].sort((a, b) => a - b),
    )
  })

  it('cada linaje aporta dos padres', () => {
    const ranking = m.rankingPadres()
    expect(ranking.reduce((suma, e) => suma + e.veces, 0)).toBe(2 * ids.length)
  })

  it('el porcentaje es coherente con las veces', () => {
    const ranking = m.rankingPadres()
    for (const entry of ranking) {
      expect(entry.porcentaje).toBeCloseTo((entry.veces * 100) / ids.length, 3)
      expect(entry.veces).toBeGreaterThanOrEqual(0)
      expect(entry.veces).toBeLessThan(ids.length)
      expect(entry.puntosMedios).toBeGreaterThanOrEqual(0)
    }
  })

  it('el orden es determinista', () => {
    const primero = m.rankingPadres()
    const segundo = m.rankingPadres()
    expect(primero).toEqual(segundo)
    for (let i = 1; i < primero.length; i++) {
      const a = primero[i - 1]
      const b = primero[i]
      if (a.veces !== b.veces) expect(a.veces).toBeGreaterThan(b.veces)
      else if (a.puntosMedios !== b.puntosMedios) expect(a.puntosMedios).toBeGreaterThan(b.puntosMedios)
      else if (a.totalAfinidad !== b.totalAfinidad) expect(a.totalAfinidad).toBeGreaterThan(b.totalAfinidad)
      else expect(a.personaje.charId).toBeLessThanOrEqual(b.personaje.charId)
    }
  })

  it('el recuento independiente coincide', () => {
    const conteo = new Map<number, number>()
    const puntos = new Map<number, number[]>()
    for (const hijoId of ids) {
      const linaje = m.mejorLinajeDe(hijoId)
      if (!linaje) continue
      for (const padreId of [linaje.padre.charId, linaje.madre.charId]) {
        conteo.set(padreId, (conteo.get(padreId) ?? 0) + 1)
        const lista = puntos.get(padreId) ?? []
        lista.push(linaje.puntos)
        puntos.set(padreId, lista)
      }
    }
    const ranking = new Map(m.rankingPadres().map((e) => [e.personaje.charId, e]))
    for (const id of ids) {
      const entry = ranking.get(id)!
      expect(entry.veces, `veces de ${id}`).toBe(conteo.get(id) ?? 0)
      const lista = puntos.get(id) ?? []
      const esperadaMedia = lista.length === 0 ? 0 : Math.trunc(lista.reduce((s, v) => s + v, 0) / lista.length)
      expect(entry.puntosMedios, `media de ${id}`).toBe(esperadaMedia)
    }
  })
})
